We begin with an Excel workbook that contains a very messy structure of data. It contains:
-Summary, where each transmission lines analysis results can be viewed one-by-one by an end user
-Impacts, where the impact of queue projects are evaluated on each transmission line
-POI List, which lists each Point of Interconnection evaluated
-Queue List, which lists each queue project evaluated
-Table Summary, which contains a table for each state with a hierarchy that shows the charge and inject capacity across lines as they are affected by different queue projects.

We also have a KML file that stores HIFLD data for each of these lines. We are interested in this data because it includes geographic features which we can map our data analysis to.

Use Case
With brevity in mind, I will conduct the analysis on a set of 3 states, which include Georgia, Alabama, and Mississipi

I will begin by making the table summaries each their own page of the workbook and deleting ones we will not use. There is little enough data that this is feasible.


USE CASE

We want to take the study done by an electrical engineering consultant and parse it to identify which transmission lines are viable for new renewable energy projects. We then want to turn these transmission lines into a map, store the data we synthesize during our analysis, and do some geographic analyses on them. We define our non-geographic viability criteria as a dictionary of lists. The dictionary's lists will be filled with values which represent arbitrary breaks in our data that we will use for comparison and analysis. 

While we start with the entire workbook, I know from browsing the tables that there are 2 in particular that we are interested in for our analyses. These are the "POI List" which contains all of the details about our transmission lines, and "Impacts", which contains the data we will use to determine viability. Let's take a look:

In [16]:
import pandas as pd
import geopandas as gpd
from keplergl import KeplerGl
import numpy as np
import fiona
fiona.drvsupport.supported_drivers['kml'] = 'rw'
fiona.drvsupport.supported_drivers['KML'] = 'rw'


In [17]:
ImpactsDF = pd.read_excel(r"C:\Users\bobro\Personal_Projects\Pterra Study - Copy.xlsx", 1)
display(ImpactsDF)

ISO Owner POI ID                 POI Description  Voltage Level Remarks  \
0    PJM    CE   IL-1             Belvidere - Marengo          138.0     NaN   
1    PJM    CE   IL-2             Marengo - Woodstock          138.0     NaN   
2    PJM    CE   IL-3          Cherry Valley - DeKalb          138.0     NaN   
3    PJM    CE   IL-4     Cherry Valley - Silver Lake          345.0     NaN   
4    PJM    CE   IL-5       Crescent Ridge - La Salle          138.0     NaN   
..   ...   ...    ...                             ...            ...     ...   
211  PJM   AEP   MI-2        New Buffalo - Three Oaks           69.0     NaN   
212  PJM   AEP   MI-3    New Carlisle - Hickory Creek          138.0     NaN   
213  PJM   AEP   MI-4             Bridgman - Pletcher           69.0     NaN   
214  PJM   AEP   MI-5  Pletcher - Buchanan South Side           69.0     NaN   
215  PJM   AEP   MI-6           Three Oaks - Lakeside           69.0     NaN   

     MW Capacity without QP  Largest Negative MW Impact  \
0                     170.0                      -170.0   
1                     130.0                      -130.0   
2                      80.0                       -40.0   
3                     390.0                      -280.0   
4                     150.0                      -150.0   
..                      ...                         ...   
211                    30.0                         0.0   
212                   150.0                      -150.0   
213                    60.0                       -10.0   
214                    70.0                       -70.0   
215                    30.0                         0.0   

    QPs with Significant Impacts (at least 30 MW reduction or caused the largest impact)  
0                   AH2-197, AG2-545, AG2-544, AF1-048                                    
1                            AH2-197, AG2-545, AG2-544                                    
2                            AH2-234, AG2-575, AG1-119                                    
3    AH2-234, AH2-197, AH2-085, AH1-657, AH1-047, A...                                    
4                            AH2-372, AG1-005, AF1-318                                    
..                                                 ...                                    
211                                                NaN                                    
212                                            AH2-373                                    
213                                            AH2-418                                    
214                                            AH2-373                                    
215                                                NaN                                    

[216 rows x 9 columns]

Here we can see some really interesting data. We have:
    POI ID
    POI Description
    Voltage Level
    MW Cap. without QP
    Largest Negative MW Impact

These will be the core of our analysis. We are interested only in economically feasible lines, which means there are criteria for MW capacity based on the Voltage level.
We also want to avoid investing resources into transmission lines where competition projects already in the queue could fill up all of the available capacity on a transmission line. The risk is that the project requires a transmission upgrade, which is an economic disaster and would likely result in an abandoned project.

For simplicity's sake, I take the largest negative MW impact as the worst case scenario. There are other possibilities that may be better or worse, but the probabilities of those scenarios are difficult to calculate and are beyond the scope of our data. Additionally, the real numbers behind the original analyses are proprietary, so I will be using arbitrary criteria as follows:

For eligibility, the largest negative MW impact on a POI may not reduce that POIs capacity below:

    >=69 kV    ----> 80 MW
    >=115 kV   ----> 100 MW
    >=138 kV   ----> 120 MW
    >=161 kV   ----> 140 MW
    >=230 kV   ----> 160 MW
    >=345 kV   ----> 180 MW

Let's see which lines make the cut:
 




In [18]:
ImpactsDF['Eval_Capacity'] = ImpactsDF['MW Capacity without QP'] + ImpactsDF['Largest Negative MW Impact']
ImpactsDF['viable'] = pd.Series
ImpactsDF['viable'] = np.where((ImpactsDF['Voltage Level'] <= 69) & (ImpactsDF['Eval_Capacity'] >= 80), True, ImpactsDF['viable'])
ImpactsDF['viable'] = np.where(((ImpactsDF['Voltage Level'] == 115) & (ImpactsDF['Eval_Capacity'] >= 100)), True, ImpactsDF['viable'])
ImpactsDF['viable'] = np.where(((ImpactsDF['Voltage Level'] == 138) & (ImpactsDF['Eval_Capacity'] >= 120)), True, ImpactsDF['viable'])
ImpactsDF['viable'] = np.where(((ImpactsDF['Voltage Level'] == 161) & (ImpactsDF['Eval_Capacity'] >= 140)), True, ImpactsDF['viable'])
ImpactsDF['viable'] = np.where(((ImpactsDF['Voltage Level'] == 230) & (ImpactsDF['Eval_Capacity'] >= 160)), True, ImpactsDF['viable'])
ImpactsDF['viable'] = np.where(((ImpactsDF['Voltage Level'] == 345) & (ImpactsDF['Eval_Capacity'] >= 180)), True, ImpactsDF['viable'])

ImpactsDF.drop(ImpactsDF[ImpactsDF.viable != True].index, inplace=True)

display(ImpactsDF[['POI ID', 'Voltage Level', 'MW Capacity without QP', 'Largest Negative MW Impact', 'Eval_Capacity', 'viable']])


POI ID  Voltage Level  MW Capacity without QP  \
7     IL-8A          138.0                   230.0   
11    IL-10          138.0                   240.0   
12    OH-1A          138.0                   390.0   
13    OH-1B          138.0                   240.0   
14     OH-2          138.0                   210.0   
15     OH-3          345.0                   600.0   
19    OH-6A          138.0                   150.0   
20    OH-6B          138.0                   200.0   
22     OH-8          138.0                   190.0   
23     OH-9          138.0                   230.0   
24    OH-10          345.0                   290.0   
25    OH-11          345.0                  1190.0   
26    OH-12           69.0                   130.0   
29   OH-14A          138.0                   170.0   
30   OH-14B          138.0                   240.0   
31   OH-14C          138.0                   340.0   
36    OH-19          138.0                   160.0   
38    OH-21           69.0                    90.0   
46    OH-29          345.0                   810.0   
50   OH-32A          138.0                   140.0   
51   OH-32B          138.0                   160.0   
60     VA-5          138.0                   190.0   
81     MS-2          115.0                   120.0   
94     WV-1          138.0                   170.0   
96     WV-3          138.0                   170.0   
104    IN-1          138.0                   280.0   
110    GA-1          115.0                   250.0   
115    GA-5          115.0                   100.0   
122   GA-11          230.0                   350.0   
123   GA-12          230.0                   310.0   
133   GA-22          230.0                   400.0   
136   GA-25          115.0                   250.0   
145   GA-34          115.0                   160.0   
146   GA-35          115.0                   160.0   
148    AL-1          161.0                   230.0   
149    AL-2          115.0                   140.0   
150    AL-3          115.0                   180.0   
159   AL-11          230.0                   500.0   
161   AL-13          161.0                   170.0   
162   AL-14          115.0                   130.0   
163   AL-15          230.0                   390.0   
165   AL-17          115.0                   250.0   
175   AL-27          230.0                   630.0   
177   AL-29          230.0                   790.0   
178   AL-30          161.0                   160.0   
183  AL-33B          161.0                   190.0   
184   AL-34          115.0                   110.0   
189    PA-3          115.0                   230.0   
192    PA-6          138.0                   150.0   
202   PA-15          115.0                   210.0   
204   PA-17          115.0                   220.0   
210    MI-1          345.0                   610.0   

     Largest Negative MW Impact  Eval_Capacity viable  
7                         -90.0          140.0   True  
11                       -120.0          120.0   True  
12                          0.0          390.0   True  
13                        -10.0          230.0   True  
14                        -70.0          140.0   True  
15                       -200.0          400.0   True  
19                        -20.0          130.0   True  
20                        -50.0          150.0   True  
22                        -20.0          170.0   True  
23                       -100.0          130.0   True  
24                        -10.0          280.0   True  
25                       -220.0          970.0   True  
26                        -40.0           90.0   True  
29                          0.0          170.0   True  
30                          0.0          240.0   True  
31                        -40.0          300.0   True  
36                          0.0          160.0   True  
38                        -10.0           80.0   True  
46                        -20.0          790.0   True  
50                    

Great! It looks like the calculations and evaluations are working right. We went from 216 rows down to 52 after dropping lines that aren't viable, and we can now take these and match them to transmission lines from HIFLDs data to generate mappable linestrings. Lets take a look at our KML.

In [19]:
KML_GDF = gpd.GeoDataFrame()
for layer in fiona.listlayers(r"C:\Users\bobro\OneDrive\Documents\Transmission Analysis Map Data.kml"):
    layer_gdf = gpd.read_file(r"C:\Users\bobro\OneDrive\Documents\Transmission Analysis Map Data.kml", driver='KML', layer=layer)
    KML_GDF = pd.concat([KML_GDF, layer_gdf], ignore_index=True)
display(KML_GDF)



Name  \
0                          Claxton - Highway 301   
1    Forstmann Louisville - North Louisville Tap   
2                  Donalsonville - Reynoldsville   
3                   Blakely - Pine Hill Junction   
4                 North Glennville - Stoney Hill   
..                                           ...   
559                                   Three Oaks   
560                                  NEW BUFFALO   
561                                         COOK   
562                                 KENZIE CREEK   
563                                Hickory Creek   

                                      Description  \
0    Transmission Line - 115kV - GEORGIA POWER CO   
1    Transmission Line - 115kV - GEORGIA POWER CO   
2    Transmission Line - 115kV - GEORGIA POWER CO   
3    Transmission Line - 115kV - GEORGIA POWER CO   
4    Transmission Line - 115kV - GEORGIA POWER CO   
..                                            ...   
559                                                 
560                                                 
561                                                 
562                                                 
563                                                 

                                              geometry  
0    MULTILINESTRING Z ((-81.85080 32.29540 40.0000...  
1    MULTILINESTRING Z ((-82.50060 32.91560 40.0000...  
2    MULTILINESTRING Z ((-84.88380 31.04300 40.0000...  
3    MULTILINESTRING Z ((-84.90600 31.37160 40.0000...  
4    MULTILINESTRING Z ((-82.03150 31.98830 40.0000...  
..                                                 ...  
559               POINT Z (-86.60633 41.80048 0.00000)  
560               POINT Z (-86.73828 41.79021 0.00000)  
561               POINT Z (-86.55417 41.97679 0.00000)  
562               POINT Z (-86.18955 41.88600 0.00000)  
563               POINT Z (-86.48106 42.07011 0.00000)  

[564 rows x 3 columns]

Awesome! It looks like "Name" attributes from the KML match up with "POI Description" from our Impacts Dataframe. Let's do a Join on that column


In [20]:
Viable_Lines_GDF = gpd.GeoDataFrame(ImpactsDF.merge(KML_GDF, how='inner', left_on='POI Description', right_on='Name'))
display(Viable_Lines_GDF)

ISO    Owner  POI ID                        POI Description  \
0    SE     SOCO    GA-1                  Claxton - Highway 301   
1    SE     SOCO    GA-5         North Glennville - Stoney Hill   
2    SE     SOCO   GA-11                            Bio - Vanna   
3    SE     SOCO   GA-12  North Tifton - Mithchell 230 Ring Bus   
4    SE     SOCO   GA-22                   Pitts - North Tifton   
5    SE     SOCO   GA-25                       Pitts - Kathleen   
6    SE     SOCO   GA-34             Claxton - North Glennville   
7    SE     SOCO   GA-35           Glenville - South Glennville   
8    SE      TVA    AL-1      Limestone - Mallard Fox Creek Tap   
9    SE       PS    AL-2                      Chastain - Repton   
10   SE     SOCO    AL-3               Flomation - South Atmore   
11   SE     SOCO   AL-11            Greene County - North Selma   
12   SE      TVA   AL-13         Mulberry Creek - Cherokee (AL)   
13   SE       PS   AL-14                 Belleville - Evergreen   
14   SE     SOCO   AL-15                Madison Park - Snowdoun   
15   SE     SOCO   AL-17          Hope Hull Tap - Burk Tap East   
16   SE     SOCO   AL-27            Snowdoun - South Montgomery   
17   SE     SOCO   AL-29  Farley Nuclear Power Plant - Pinckard   
18   SE      TVA   AL-30                Wheeler Dam - Ford City   
19   SE      TVA  AL-33B                   Wilson Dam - Oakland   
20   SE       PS   AL-34    Opine Road - Bridges Creek Junction   
21  PJM  PENELEC    PA-3           Eerie South - Edinboro South   
22  PJM  PENELEC   PA-15          Venango Junction - Springboro   
23  PJM  PENELEC   PA-17            Erie Southeast - Corry East   
24  PJM      AEP    MI-1                    Cook - Kenzie Creek   

    Voltage Level Remarks  MW Capacity without QP  Largest Negative MW Impact  \
0           115.0     NaN                   250.0                       -20.0   
1           115.0     NaN                   100.0                         0.0   
2           230.0     NaN                   350.0                         0.0   
3           230.0     NaN                   310.0                      -130.0   
4           230.0     NaN                   400.0                      -190.0   
5           115.0     NaN                   250.0                       -90.0   
6           115.0     NaN                   160.0                       -30.0   
7           115.0     NaN                   160.0                       -10.0   
8           161.0     NaN                   230.0                       -30.0   
9           115.0     NaN                   140.0                         0.0   
10          115.0     NaN                   180.0                       -70.0   
11          230.0     NaN                   500.0                      -300.0   
12          161.0     NaN                   170.0                         0.0   
13          115.0     NaN                   130.0                         0.0   
14          230.0     NaN                   390.0                       -20.0   
15          115.0     NaN                   250.0                       -80.0   
16          230.0     NaN                   630.0                       -70.0   
17          230.0     NaN                   790.0                      -150.0   
18          161.0     NaN                   160.0                       -10.0   
19          161.0     NaN                   190.0                         0.0   
20          115.0     NaN                   110.0                       -10.0   
21          115.0     NaN                   230.0                      -100.0   
22          115.0     NaN                   210.0                       -50.0   
23          115.0     NaN                   220.0                       -70.0   
24          345.0     NaN                   610.0                       -80.0   

   QPs with Significant Impacts (at least 30 MW reduction or caused the largest impact)  \
0                                              IC-625          

Viable_Lines_Errors = ImpactsDF.merge(KML_GDF, how='left', left_on='POI Description', right_on='Name')
ImpactsDF.drop(ImpactsDF[ImpactsDF.viable != True].index, inplace=True)

In [21]:
Viable_Lines_Errors = ImpactsDF.merge(KML_GDF, how='left', left_on='POI Description', right_on='Name', indicator='source')
Viable_Lines_Errors.drop(Viable_Lines_Errors[Viable_Lines_Errors.geometry != None].index, inplace=True)
display(Viable_Lines_Errors)

ISO Owner  POI ID                  POI Description  Voltage Level Remarks  \
0   PJM    CE   IL-8A             Rock Falls - Kewanee          138.0     NaN   
1   PJM    CE   IL-10               Steward - Waterman          138.0     NaN   
2   PJM   AEP   OH-1A    South East Canton - Blue Bell          138.0     NaN   
3   PJM   AEP   OH-1B    South East Canton - Blue Bell          138.0     NaN   
4   PJM  ATSI    OH-2               Toussaint - Ottawa          138.0     NaN   
5   PJM  ATSI    OH-3             Davis Besse - Beaver          345.0     NaN   
6   PJM  ATSI   OH-6A           Galion - Dual Rail Tap          138.0     NaN   
7   PJM  ATSI   OH-6B           Galion - Dual Rail Tap          138.0     NaN   
8   PJM  ATSI    OH-8     East Springfield - Millcreek          138.0     NaN   
9   PJM  ATSI    OH-9    East Springfield - London Tap          138.0     NaN   
10  PJM   DAY   OH-10         West Milton - Miami Fort          345.0     NaN   
11  PJM   DAY   OH-11                   Greene - Beaty          345.0     NaN   
12  PJM   DAY   OH-12            Atlanta - New Holland           69.0     NaN   
13  PJM   AEP  OH-14A       Reedsburgh - Dilles Bottom          138.0     NaN   
14  PJM   AEP  OH-14B       Reedsburgh - Dilles Bottom          138.0     NaN   
15  PJM   AEP  OH-14C       Reedsburgh - Dilles Bottom          138.0     NaN   
16  PJM   AEP   OH-19              Wolf Creek - Layman          138.0     NaN   
17  PJM   DAY   OH-21              Rossburg - Rosehill           69.0     NaN   
18  PJM   DAY   OH-29            J.M. Stuart - Clinton          345.0     NaN   
19  PJM  ATSI  OH-32A                    Camden - Troy          138.0     NaN   
20  PJM  ATSI  OH-32B                    Camden - Troy          138.0     NaN   
21  PJM   APS    VA-5      Double Toll Gate - Riverton          138.0     NaN   
22   SE  SOCO    MS-2      Lumberton - Hurricane Creek          115.0     NaN   
23  PJM   APS    WV-1           Belington - Union Road          138.0     NaN   
24  PJM   APS    WV-3              Pruntytown - Shinns          138.0     NaN   
25  PJM   AEP    IN-1             Dunton Lake - Varner          138.0     NaN   
48  PJM   APS    PA-6  Connellsville - Stoner Junction          138.0     NaN   

    MW Capacity without QP  Largest Negative MW Impact  \
0                    230.0                       -90.0   
1                    240.0                      -120.0   
2                    390.0                         0.0   
3                    240.0                       -10.0   
4                    210.0                       -70.0   
5                    600.0                      -200.0   
6                    150.0                       -20.0   
7                    200.0                       -50.0   
8                    190.0                       -20.0   
9                    230.0                      -100.0   
10                   290.0                       -10.0   
11                  1190.0                      -220.0   
12                   130.0                       -40.0   
13                   170.0                         0.0   
14                   240.0                         0.0   
15                   340.0                       -40.0   
16                   160.0                         0.0   
17                    90.0                       -10.0   
18                   810.0                       -20.0   
19                   140.0                       -10.0   
20                   160.0                         0.0   
21                   190.0                       -20.0   
22                   120.0                       -10.0   
23                   170.0                       -40.0   
24                   170.0                         0.0   
25                   280.0                       -70.0   
48                   150.0                         0.0   

   QPs with Significant Impacts (at least 30 MW reduction or caused the largest impact)  \
0   AH2-376, AH2-37

Excellent! I'll review these ones and match up the best ones with lines on the KML using their locations and descriptions.

For now, let's take a look at the map we've got so far:

In [22]:
Viable_Lines_GDF = Viable_Lines_GDF.explode()

map = KeplerGl()
map.add_data(data=Viable_Lines_GDF, name='Viable Transmission Lines')
map

User Guide: https://docs.kepler.gl/docs/keplergl-jupyter


c:\Users\bobro\anaconda3\envs\Transmission_Analysis\lib\site-packages\jupyter_client\session.py:718: UserWarning: Message serialization failed with:
Out of range float values are not JSON compliant
Supporting this message is deprecated in jupyter-client 7, please make sure your message is JSON-compliant
  content = self.pack(content)


KeplerGl(data={'Viable Transmission Lines': {'index': [(0, 0), (1, 0), (2, 0), (3, 0), (4, 0), (5, 0), (6, 0),…